File: tsp_dynprog.pl
use strict;
use warnings;
use List::Util qw(min);

my @adj_matrix;
my $start_city;

# Read the adjacency matrix from a file
sub get_matrix_from_file {
    my ($file_path) = @_;
    
    open my $file, '<', $file_path or die "Could not open file: $!";
    while (my $line = <$file>) {
        chomp $line;
        push @adj_matrix, [split(/\s+/, $line)];
    }
    close $file;
}

#Checking if the matrix valid
sub is_matrix_valid {
    my ($matrix) = @_;
    my $num_rows = @$matrix;
    return 0 unless $num_rows > 0;  

    my $num_cols = @{$matrix->[0]};
    foreach my $row (@$matrix) {
        return 0 unless @$row == $num_cols; 
    }

    return 1;
}

# Solve the TSP using dynamic programming
sub tsp_with_dp {
    my ($i, @S) = @_;
    
    # Base case: if S has only 1 item left, return distance from it to the starting point
    if (scalar(@S) == 1) {
        my $j = $S[0];
        return ([$i, $j, $start_city], $adj_matrix[$i][$j] + $adj_matrix[$j][$start_city]);
    }
    
    #Declaring minimum distance as a very large number
    my @result = (undef, 10**9);
    
    for my $j (@S) {
        my @new_S = grep { $_ != $j } @S;
        
        # Recursion
        my ($path, $cost) = tsp_with_dp($j, @new_S);
        unshift @$path, $i;  # Insert source to the result array
        my $recursion_result = ([$path, $cost + $adj_matrix[$i][$j]]);
        @result = ($result[1] > $recursion_result->[1]) ? @$recursion_result : @result;
    }
    
    return @result;
}

# -------------------- MAIN PROGRAM -------------------------
print "Hi, salesman! Load the input matrix file of your map. Write 'input_matrix_X.txt': ";
chomp(my $input_matrix_file = <STDIN>);
get_matrix_from_file($input_matrix_file);
if (!is_matrix_valid(\@adj_matrix)){
    die "The matrix is not valid.\n";
}

my $num_of_cities = scalar @{$adj_matrix[0]};
print "There are ", $num_of_cities, " cities.\n";
print "Hi, salesman! Which city are you going from? (Index from 0) ";
chomp($start_city = <STDIN>);
if ($start_city < 1 || $start_city > $num_of_cities) {
    die "The value of ($start_city) is out of the range 1 to $num_of_cities.\n";
}
$start_city -= 1;

my @points_of_cities = grep { $_ != $start_city } 0..$#{$adj_matrix[0]};

my ($path, $cost) = tsp_with_dp($start_city, @points_of_cities);
print "Most optimum TSP route using Dynamic Programming: ";
my @incremented_path = map { $_ + 1 } @$path;
print join(" ", @incremented_path), "\n";
print "Approximate distance to travel: $cost\n";

File: tsp_dynprog.py
import math

def get_matrix_from_file(file_path):
    matrix = []
    with open(file_path, "r") as file:
        for line in file:
            elements = line.strip().split()
            matrix.append([int(element) for element in elements])
    return matrix


'''
Find optimum solution for Travelling Salesman Problem using Dynamic Programming
@param i : source
@param S : set of destinations 
'''
def tsp_with_dp(i, S):
    ''' Base : if S has only 1 item left, return distance from it to the ending point '''
    if (len(S) == 1):
        j = S.pop()
        return ([i, j, start_city],  adj_mtrx[i][j] + adj_mtrx[j][start_city])
 
    # Set default value for the result
    result = ([], 10**9)

    # Iterate through all item in S
    for j in S:

        ''' Recursion '''
        recursion = tsp_with_dp(j, S-{j})

        # Insert source to the result array
        recursion[0].insert(0,i)

        # Create new tuple for the result and add distance from source(i) to current destination (j)
        recursion_result = (recursion[0], recursion[1] + adj_mtrx[i][j])
    
        # Take result only if minimum, compared by the distance
        result = min(result, recursion_result, key=lambda x: x[1])

    return result


adj_mtrx = get_matrix_from_file("input_matrix.txt")
print(f"There are {len(adj_mtrx[0])} cities.")
start_city = int(input("Hi, salesman! Which cities you are going from? (Index from 0) "))
# Create set of coordinates index to evaluate
points_of_cities = set(range(0, len(adj_mtrx[0]))) - {start_city}

# Find solution using Dynamic Programming
tsp_with_dp_result = tsp_with_dp(start_city, points_of_cities)
print("Most optimum TSP route using Dynamic Programming: ", end=" ")
for place in tsp_with_dp_result[0]:
    print(place, end=" ")
print(f"\nApproximate distance to travel : {tsp_with_dp_result[1]}")



